### DISCUSSION TITLE
Can Flutter Web using FRB be used in HTTP only networks?

### DISCUSSION THREAD

**Discussion by CJBuchel** [fzyzcjy/flutter_rust_bridge#2401] [Q&A] [answered]
I'm trying to get a local web server setup using the rust warp crate and flutter compiled for the web using FRB WASM.
When it comes to the web my understanding is that FRB uses some features that require specific headers for cross origin isolation https://cjycode.com/flutter_rust_bridge/manual/miscellaneous/web-cross-origin

I'm not certain what these features are, but from a small amount of research. It might be for shared buffer arrays?
But in doing so I've often run into issues just trying to simply host the webserver, which I think has to do with my browser (most browsers) enforcing that those specific headers can only work over https for "security concers" :(

I have been able to get it working since using https with self signed certificates, but it's not ideal. And causes a lot of issues when I host the webserver in an environment where the security is setup to block self signed certificates as well. (Sadly I can't avoid the self signed certs).

Is there any way to use FRB with just http? And not require the webserver to be on https?
I'm not too familar all round with the headers and the security of them, but I was under the impression that WASM could be used over http. So are the features used in FRB possibly optional?

Thanks in advance :)

**Comment by fzyzcjy** [comment]
Firstly, may I know your usage scenario? It is recommended to have HTTPS when hosting a website, otherwise it can be pretty dangerous and have attacks.

Secondly, yes it is because some requirements in Rust wasm. But if you never use things outside of the main thread, then I guess it may be possible to remove that header. But again I would suggest firstly have a discussion about why HTTPS cannot be deployed.

**Reply by CJBuchel** [reply]
The usecase is very niche.
https://github.com/CJBuchel/TMS

I have a project designed for one day deployments for small competitions hosted at schools and stadiums. Where in most cases you either take a portable router with you to host the webserver on, or in best cases use the school provided wifi. And just launch the application on a laptop or something similar for the day.

Then you connect up ipads or browsers to input and display information before eventually exporting it all. And shutting it down.
It makes things fairly easy because live scoreboards and timers can be projected on the other side of the campus using existing infrastructure which schools and stadiums usually already offer. Just an HDMI cable and a simple connection to the local network and we're done.

So security wise, it's not much of an issue as it's internal to just the school network. For a single day, and if needed. I can usually build up custom encryption to keep learing eyes away if abolultely necessary. But, I've never run into an issue.

At the moment I do use https, but I'll often run into an issue with the network just outright blocking it because the certificate is self signed. I can't exactly get a proper certificate either because it's hosted on simple local networks which change for each event. And asking the venue to shutdown their firewall is usually a no go, as it runs on the weekend and it's a pain for them. Usually though I run into this issue more often because of a school specific device, like one of their own laptops or ipads. Which are configured to not accept the self signed certs. So they simply don't allow the user to even say "I understand, continue anyway".

It's very niche, and usability in this case is more of a priority than security. So I've been trying to find ways to mitigate this as much as possible, but I can't fix every possible browser security enforcements. And every school or venue firewal configs.
So having a way to use FRB over http would be a great help. (My preference is that my application supports both, and by default uses https if it can. But the user can switch down to http if the venue is a pain)

**Reply by fzyzcjy** [reply]
Ah I see! Let me think about it.

Btw I knew "FTC" (FIRST tech challenge) when in high school. Looks like your repo is somehow related to FIRST?

**Reply by fzyzcjy** [reply]
Workarounds:

* Is it possible to host your website on a real domain (instead of local network)? (Shortcomings: Larger latency, need to pay for servers, etc)
* Is it possible to let users download app instead of using webpage? (Shortcomings: users may be unhappy)

Solutions:

Try to restrict ourselves to only use main thread and never use other threads. Then we can run without that header.

* `pub async fn`: On main thread (when in web)
* `#[frb(sync)] pub fn`: On main thread
* `pub fn`: NOT on main thread by default

**Reply by CJBuchel** [reply]
Ah, what are the chances. I'm actually helping out with an FTC event tomrrow, first time in a while in WA. Should be fun.
Yep, Repo is for FLL to score and display results.

**Reply by fzyzcjy** [reply]
Just curious, do you mean official FLL competitions or not? (I guess it is) What are the chances - one day my code can be used in FIRST's competitions that I knew long ago!

Btw, I blurly remembered someone long ago may ask similar things of trying to run *only* on main thread on web. Let me see if I can find that issue.
EDIT: Found it https://github.com/fzyzcjy/flutter_rust_bridge/issues/2215#issuecomment-2259967958

**Reply by fzyzcjy** [reply, marked as answer]
Then looks like your case can be done, given #2215 seemed to succeed. Also feel free to PR for a quick doc change if you like!

**Reply by fzyzcjy** [reply]
Oh I see it: https://www.firstaustralia.org/all-events

![image](https://github.com/user-attachments/assets/cd7f9c19-f45f-403f-bc29-450e185591b6)

Have a nice day :)

**Reply by CJBuchel** [reply]
Sadly using an actual domain goes against the main reason this project was started. In Aus we often have some comps that just don't have internet at all in the outback. So a small router and a case full of iPads is toured around the outbacks, and we make do with bringing as much "flare" as we can.

I'm more inclined to use single threads only. I very rarely use async anyway with FRB. I mostly use it for shared structures between server and client. And I also use it to serialize the data to json using server side code. Which I want to switch out anyway to a protobuf solution with a more standardised network. So only a small portion of logic is really shared between the server and client. None of it needing to be async.

"side notes"
- Yep, technically official events. Never really thought about it like that XD. Was just a bit of fun on the side to make things a bit cooler for the kids.

**Reply by CJBuchel** [reply]
Thanks, I'll have a look at setting it to http.

**Reply by fzyzcjy** [reply]
You are welcome!

Btw try `default_dart_async: false` to avoid annotating each and every function with `#[frb(sync)]`

**Reply by CJBuchel** [reply]
Ah yep, will do.
